### PR TITLE
Dynamic hp models

### PR DESCRIPTION
**Pull request by cwastche**
### **PR Type**
Enhancement

___

### **Description**
- Added dynamic loading of multiple model paths for structures in `StructureManager`.
- Introduced progress-based model selection for Hyperstructures.
- Updated structure preview to handle multiple model paths.
- Refactored structure coordinates mapping and loading logic.
- Adjusted realm entity ID initialization in `Worldmap` scene.
- Defined progress thresholds and enums for structure progress in constants.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>StructureManager.ts</strong><dd><code>Dynamic model loading and progress-based selection for structures</code></dd></summary>
<hr>

client/src/three/components/StructureManager.ts

<li>Added dynamic loading of multiple model paths for structures.<br> <li> Introduced progress-based model selection for Hyperstructures.<br> <li> Updated structure management to handle multiple models per structure <br>type.<br> <li> Refactored structure coordinates mapping and loading logic.<br>

</details>

  </td>
  <td><a href="https://github.com/BibliothecaDAO/eternum/pull/1228/files#diff-84ab022ecfbae90e696e83a20edd931ed1c82a497f7f454b0b517f76d8c9eb7f">+130/-46</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>StructurePreview.ts</strong><dd><code>Enhanced structure preview with multiple model paths</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

client/src/three/components/StructurePreview.ts

<li>Updated structure preview to load multiple model paths.<br> <li> Enhanced error handling for model loading.<br>

</details>

  </td>
  <td><a href="https://github.com/BibliothecaDAO/eternum/pull/1228/files#diff-635f573840762cced0c89f6441d862caa3149526d33fd9a2dfdc8f5a2079f3d2">+26/-25</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Worldmap.ts</strong><dd><code>Refactor Worldmap scene for enhanced structure management</code></dd></summary>
<hr>

client/src/three/scenes/Worldmap.ts

<li>Updated <code>StructureManager</code> instantiation to include <code>dojo</code>.<br> <li> Refactored structure coordinates mapping.<br> <li> Adjusted realm entity ID initialization.<br>

</details>

  </td>
  <td><a href="https://github.com/BibliothecaDAO/eternum/pull/1228/files#diff-39f176e11a1d117525a8de19ae6404d20179a1d8959de6b86b17a1aa9f2de4ad">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>constants.ts</strong><dd><code>Define multiple model paths and progress thresholds for structures</code></dd></summary>
<hr>

client/src/three/scenes/constants.ts

<li>Defined multiple model paths for structures.<br> <li> Introduced progress thresholds and enums for structure progress.<br>

</details>

  </td>
  <td><a href="https://github.com/BibliothecaDAO/eternum/pull/1228/files#diff-1f49c55dac9ff065a9c6f5456b4c6bb0577d190dc08d19d8266719e7c91c4425">+22/-11</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: client/src/three/components/StructureManager.ts [hunk 1]
```diff
@@ -88,18 +120,73 @@ export class StructureManager {
...
+
+        // Add label on top of the structure with appropriate color
+        const labelColor = isMine ? myColor : neutralColor;
+        const label = this.labelManagers.get(key)?.createLabel(position as any, labelColor);
+        this.scene.add(label!);
+      }
     }
   }
+
+  getModelIndex(structureType: StructureType, entityId: ID): number {
```
Comment (aymericdelab) [inline client/src/three/components/StructureManager.ts:139]: in ArmyManager and StructureManager we made sure to only handle model stuff and not have any dojo and recs there

Comment (aymericdelab) [reply client/src/three/components/StructureManager.ts:139]: all recs is managed in SystemManager

Comment (aymericdelab) [inline client/src/three/components/StructureManager.ts:139]: getModelIndex is just for Hyperstructure progress? if so name should be more explicit

File: client/src/three/components/StructurePreview.ts [hunk 2]
```diff
@@ -15,33 +15,34 @@ export class StructurePreview {
...
+                  child.material.opacity = 0.75;
+                }
+              });
+              this.structureModels.set(parseInt(building), model);
+              resolve();
+            },
+            undefined,
+            (error) => {
+              console.error(`An error occurred while loading the ${[building]} model:`, error);
+              reject(error);
```
Comment (edisontim) [inline client/src/three/components/StructurePreview.ts:39]: Could you put all of this inside a function to avoid different levels of abstraction?

**Review by edisontim — APPROVED** [review]

**Review by edisontim — APPROVED** [review]

**Comment by edisontim** [comment]
Just needs a rebase, otherwise looks good

File: client/src/three/components/StructureManager.ts [hunk 3]
```diff
@@ -7,78 +7,102 @@ import { StructureSystemUpdate } from "../systems/types";
...
+import { ClientComponents } from "@/dojo/createClientComponents";
+import { SetupResult } from "@/dojo/setup";
 
 const neutralColor = new THREE.Color(0xffffff);
 const myColor = new THREE.Color("lime");
 
 const MAX_INSTANCES = 1000;
 
 export class StructureManager {
+  private models: {
```
Comment (aymericdelab) [inline client/src/three/components/StructureManager.ts:20]: remove the dojo models

**Pull request merged**